Descobrindo a senha
===================



## Objetivo



Usar um algoritmo genético para descobrir uma senha.



## Descrição do problema



Neste problema, a função objetivo deve saber a senha correta e quantificar de alguma maneira o quão perto ou longe os palpites estão da solução (veja que isso é algo que não temos no mundo real. Nenhum site irá te dizer se você está acertando ou errando seu palpite). O critério de parada deste problema é quando a senha for descoberta.



## Importações



In [3]:
from funcoes import populacao_inicial_senha
from funcoes import funcao_objetivo_pop_senha
from funcoes import selecao_torneio_min
from funcoes import cruzamento_ponto_simples as funcao_cruzamento
from funcoes import mutacao_senha
import random

## Códigos e discussão



In [66]:
### CONSTANTES

# relacionadas à busca
TAMANHO_POP = 50
#NUM_GERACOES = 2000
CHANCE_CRUZAMENTO = 0.5
CHANCE_MUTACAO = 0.05
NUM_COMBATENTES_NO_TORNEIO = 3

# relacionadas ao problema a ser resulvido
#SENHA = "correcthorsebatterystaple"
SENHA = "Corinthians X Cruzeiro dia 16/04"
LETRAS_POSSIVEIS = "abcdefghijklmnopqrstuvwxyzZYXC0123456789 /"
NUM_GENES = len(SENHA)

In [67]:
# funções locais

def cria_populacao_inicial(tamanho, tamanho_senha):
    return populacao_inicial_senha(tamanho, tamanho_senha, LETRAS_POSSIVEIS)

def funcao_objetivo_pop(populacao):
    return funcao_objetivo_pop_senha(populacao, SENHA)

def funcao_selecao(populacao, fitness):
    return selecao_torneio_min(populacao, fitness, NUM_COMBATENTES_NO_TORNEIO)

def funcao_mutacao(individuo):
    return mutacao_senha(individuo, LETRAS_POSSIVEIS)

In [68]:
populacao = cria_populacao_inicial(TAMANHO_POP, NUM_GENES)

melhor_fitness_ja_visto = float("inf")  # é assim que escrevemos infinito em python

print("Progresso da melhor senha já vista:")

#for n in range(NUM_GERACOES):

while melhor_fitness_ja_visto !=0:    
    
    # Seleção
    fitness = funcao_objetivo_pop(populacao)
    populacao = funcao_selecao(populacao, fitness)
    
    # Cruzamento
    pais = populacao[0::2]
    maes = populacao[1::2]
    
    contador = 0
    
    for pai, mae in zip(pais, maes):
        if random.random() <= CHANCE_CRUZAMENTO:
            filho1, filho2 = funcao_cruzamento(pai, mae)
            populacao[contador] = filho1
            populacao[contador + 1] = filho2
        
        contador = contador + 2   
        
    # Mutação
    for n in range(len(populacao)):
        if random.random() <= CHANCE_MUTACAO:
            individuo = populacao[n]
            populacao[n] = funcao_mutacao(individuo)            
            
    # melhor individuo já visto até agora
    fitness = funcao_objetivo_pop(populacao)
    menor_fitness = min(fitness)
    if menor_fitness < melhor_fitness_ja_visto:        
        posicao = fitness.index(menor_fitness)
        melhor_individuo_ja_visto = populacao[posicao]
        melhor_fitness_ja_visto = menor_fitness
        print("".join(melhor_individuo_ja_visto), "- fitness:", melhor_fitness_ja_visto)

print()
print("Melhor palpite da senha encontrado:")
print("".join(melhor_individuo_ja_visto))

Progresso da melhor senha já vista:
 ntzgcaubpgwjuCZakqpszqzkh7k431k - fitness: 648
 ntzgcaumxshj/2n2mqpszqzkh7k431k - fitness: 612
6diXekwklxshj/2nww5pszqzkh7k431k - fitness: 559
6diXekwklxshj/2nww5eXqbzkh7k431k - fitness: 557
9Ytzgcaumxshj/2natqpszqzkh7k431k - fitness: 554
6diopcaumxshj/2natqpszqzkh7k431k - fitness: 537
6diXekwklxshj/2nww5eXq zkh7k431k - fitness: 491
6diXekwklxshj/2nww5eXq zkh74431k - fitness: 436
6dtzgcakmgsqj/Cnww5eXq zkh74431k - fitness: 417
6niopcaumxshj/2natqeXq zkh74431k - fitness: 404
6dtzgcakmgsqj/CnwwqeXq zkh74431k - fitness: 381
6niogcaumxshj/2natqeXq zkh744314 - fitness: 354
6niogcaumxshj/2natqehq zkh744314 - fitness: 338
6niopcaumgsqj/CnwwqeXq zkh744314 - fitness: 317
6dtopcaummshj/CnwwqeXq zkh744314 - fitness: 305
6niopuaumgsqj/CnwwqeXq zkh744314 - fitness: 301
6niopuaumgsqj/CnwwqeXq vkh744314 - fitness: 297
6niopuakmgsqj/CnwwqeXq zkh744314 - fitness: 291
6niopuaummshj/CnwwqeXq zkh744314 - fitness: 286
6niopuakmmshj/CnwwqeXq zkh744314 - fitness: 276
6nio

## Conclusão



O desafio desse experimento é saber a senha correta e quantificar de alguma maneira o quão perto ou longe os palpites estão da solução, desse modo, o resultado foi alcançado e diferente dos demais experimento, o critério de parada desse não dependia do número de gerações e sim até a senha correta ser determinada. Portanto, definindo novas funções e dando ênfase a função "selecao_torneio_min", essa que possibilitou determinar o quão próximo um palpite está da resposta correta. E nesse experimento, a senha correta é (Corinthians X Cruzeiro dia 16/04) considerando alguns caracteres além de letras para que essa senha se torne possível, sendo que foi necessária 648 tentativas quando rodei. Contudo, esse valor não é definido, mas sempre resultara na senha correta devido ao critério de parada. Observação: essa senha é um convite.

## Playground

